### PR TITLE
Typed Builder implementation part 13.5: Standardized `ScopedBuilder.decl` tests

### LINKED ISSUE #1980: Standardize declaration tests in `HybridTests`
Follow-up to #1969

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] Entries added to [./unreleased/][unreleased] for any new functionality

closes #1980 

[unreleased]: https://github.com/informalsystems/apalache/tree/unstable/.unreleased

### CODE REVIEW

**Review by konnov — COMMENTED** [review]
I looked through the code. But I am not sure what is going on here. The issue is also not explaining much. Could you maybe add a few sentence that explain the purpose of this refactoring?

**Comment by Kukovec** [comment]
> I looked through the code. But I am not sure what is going on here. The issue is also not explaining much. Could you maybe add a few sentence that explain the purpose of this refactoring?

https://github.com/informalsystems/apalache/issues/1969 generalized the `checkRun` framework, so that it can handle builder methods that produce not just `TBuilderInstruction` (i.e. `TBuilderInternalState[TlaEx]`), but any `TBuilderInternalState[T]` (in particular, `T = TlaOperDecl`). 

The original tests for `ScopedBuilder.decl` were written before this was the case, and so were significantly different (in form) from the rest of the builder tests. This PR brings them more in line and makes the shape of the `decl` tests analogous to that of the other builder method tests.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala [hunk 1]
```diff
@@ -55,131 +57,164 @@ class TestHybrid extends BuilderTest {
...
   test("decl") {
 
-    type T = TBuilderOperDeclInstruction
+    type ExplicitT = (String, TBuilderInstruction, Seq[TypedParam])
     type TParam = (TlaType1, Seq[TlaType1])
 
-    // A(p1,...,pn) == CHOOSE x: p1 /\  p2 >= 0 /\ ... /\ pn = pn
-    def mkWellTyped(tparam: TParam): (T, T) = {
+    // We create a different term for operator vs simple parameters
+    def paramCond(param: TypedParam): TBuilderInstruction = {
```
Comment (thpani) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:64]: This needs more explanation.
Apparently it takes operators and simple parameters, but I have no idea what it's supposed to return or how it's supposed to be used.

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala [hunk 2]
```diff
@@ -55,131 +57,172 @@ class TestHybrid extends BuilderTest {
...
   test("decl") {
 
-    type T = TBuilderOperDeclInstruction
+    type ExplicitT = (String, TBuilderInstruction, Seq[TypedParam])
     type TParam = (TlaType1, Seq[TlaType1])
 
-    // A(p1,...,pn) == CHOOSE x: p1 /\  p2 >= 0 /\ ... /\ pn = pn
-    def mkWellTyped(tparam: TParam): (T, T) = {
+    // To test the declaration constructor, we need to test the relation between declared parameters, and the
+    // way the parameters ar actually used in the declaration body. Concretely, we need to create a declaration body,
```
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:64]:
```suggestion
    // way the parameters are actually used in the declaration body. Concretely, we need to create a declaration body,
```

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala [hunk 3]
```diff
@@ -55,131 +57,172 @@ class TestHybrid extends BuilderTest {
...
+    // in which every parameter p, with expected type T appears in an expression that only permits values of type T.
```
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:65]:
```suggestion
    // in which every parameter p, with expected type T, appears in an expression that only permits values of type T.
```

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala [hunk 4]
```diff
@@ -55,131 +57,172 @@ class TestHybrid extends BuilderTest {
...
+    // This way, we can test whether the builder correctly detects a discrepancy between p declared with type T
+    // and p used with type TT (s.t. TT /= T).
+    //
+    // We could always just do \E (q: declaredTypeOf(p)): q = p, but in real specs, operator types can't appear like that
+    // To test operator parameters with types (a1,...,an) => b, we construct values (q1: a1, ..., qn: an) and (e: b),
+    // then test application. For non-operator parameters, we just test \E (q: typeOf(p)): q = p
+    def paramCond(param: TypedParam): TBuilderInstruction = {
+      val (OperParam(pName, _), tt) = param
+      def mkNameTT(s: String) = builder.name(s, tt)
```
Comment (shonfeder) [inline tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:74]: just a matter of taste here, but I don't find this abstraction helps with readability.

**Review by shonfeder — APPROVED** [review]
Just a few, purely optional superficial suggestions and a typo or fix or two from me.

Looks good!

File: tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala [hunk 5]
```diff
@@ -55,131 +57,164 @@ class TestHybrid extends BuilderTest {
...
     type TParam = (TlaType1, Seq[TlaType1])
 
-    // A(p1,...,pn) == CHOOSE x: p1 /\  p2 >= 0 /\ ... /\ pn = pn
-    def mkWellTyped(tparam: TParam): (T, T) = {
+    // We create a different term for operator vs simple parameters
+    def paramCond(param: TypedParam): TBuilderInstruction = {
+      val (OperParam(pName, _), tt) = param
+      def mkNameTT(s: String) = builder.name(s, tt)
+      tt match {
+        // an operator parameter p(_,...,_) gets the term p(v1_p, ...., vn_p) = e_p
```
[the remark this answers is not in the document]
Comment (Kukovec) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:76]: `paramCond(p(_,...,_))` returns an expression of the form `p(v1_p, ...., vn_p) = e_p`

[the remark this answers is not in the document]
Comment (shonfeder) [reply tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/TestHybrid.scala:76]: If the comment said that I’d find it much clearer :)

**Review by thpani — APPROVED** [review]

**Pull request merged**